GPUP: WebCodecVideoFrame of ImageBitmap should be constructed via NativeImage

kkinnunen-apple opened the pull request:
#### 3bd96b9654855318a609f1fcb085c8bd49f186f7
<pre>
GPUP: WebCodecVideoFrame of ImageBitmap should be constructed via NativeImage
<a href="https://bugs.webkit.org/show_bug.cgi?id=301925">https://bugs.webkit.org/show_bug.cgi?id=301925</a>
<a href="https://rdar.apple.com/164004657">rdar://164004657</a>

Reviewed by NOBODY (OOPS!).

WebCodecVideoFrame of ImageBitmap would be constructed with conversion
path ImageBitmap -&gt; ImageBuffer -&gt; PixelBuffer -&gt; VideoFrame.

Use path of ImageBitmap -&gt; ImageBuffer -&gt; NativeImage -&gt; VideoFrame
instead. This gets it closer to the eventual target of ImageBitmap
-&gt; NativeImage -&gt; VideoFrame.

This is work towards being able to hold NativeImage in ImageBitmap,
ImageBitmap -&gt; NativeImage -&gt; VideoFrame. This is performance and
correctness target, where ImageBitmap should not do redundant draws
of image sources. ImageBuffer is a bitmap draw target, where as
NativeImage is a bitmap image source.

Later on, VideoFrame from NativeImage code will be able to construct
the VideoFrame directly out of GPUP-side NativeImage, avoiding
WCP-&gt;GPUP transfers.

* Source/WebCore/Modules/webcodecs/WebCodecsVideoFrame.cpp:
(WebCore::WebCodecsVideoFrame::create):
(WebCore::WebCodecsVideoFrame::initializeFrameWithResourceAndSize):
* Source/WebCore/Modules/webcodecs/WebCodecsVideoFrame.h:
* Source/WebCore/html/ImageBitmap.cpp:
(WebCore::ImageBitmap::copyNativeImage):
* Source/WebCore/html/ImageBitmap.h:
* Source/WebCore/platform/VideoFrame.cpp:
(WebCore::VideoFrame::createFromNativeImage):
(WebCore::VideoFrame::fromNativeImage): Deleted.
* Source/WebCore/platform/VideoFrame.h:
* Source/WebCore/platform/graphics/cv/VideoFrameCV.mm:
(WebCore::VideoFrame::createFromNativeImage):
(WebCore::VideoFrame::fromNativeImage): Deleted.
* Source/WebCore/platform/graphics/gstreamer/VideoFrameGStreamer.cpp:
(WebCore::VideoFrame::createFromNativeImage):
(WebCore::VideoFrame::fromNativeImage): Deleted.
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3bd96b9654855318a609f1fcb085c8bd49f186f7

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/130431 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/2702 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/41385 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/137849 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/82024 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/843d9427-4327-4ca9-b7b4-70f86a315037) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/132302 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/2708 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/2594 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/99370 "Found 27 new test failures: http/wpt/webcodecs/encoder-task-failing.html http/wpt/webcodecs/videoFrame-colorSpace.html http/wpt/webcodecs/videoFrame-duration.html http/wpt/webcodecs/videoFrame-negative-timestamp.html http/wpt/webcodecs/webcodecs-gc.html imported/w3c/web-platform-tests/webcodecs/full-cycle-test.https.any.html?h264_annexb imported/w3c/web-platform-tests/webcodecs/full-cycle-test.https.any.html?h265_annexb imported/w3c/web-platform-tests/webcodecs/full-cycle-test.https.any.html?h265_hevc imported/w3c/web-platform-tests/webcodecs/full-cycle-test.https.any.worker.html?h264_annexb imported/w3c/web-platform-tests/webcodecs/full-cycle-test.https.any.worker.html?h265_annexb ... (failure)") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/67229 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/8c6fbcaa-2caa-42a5-a5f4-de1b9e463200) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/133378 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/1977 "Found 60 new test failures: editing/selection/ios/become-first-responder-after-relinquishing-focus.html fast/forms/file/file-input-reset-using-open-panel.html fast/forms/ios/file-upload-panel-blur-focus-document-picker.html fast/media/template-hidden.html fast/mediasession/metadata/artworkdownload-svg.html fast/mediasession/metadata/artworkdownload.html fast/mediastream/MediaStream-MediaElement-setObject-null.html fast/mediastream/MediaStream-removeTrack-while-playing.html fast/mediastream/MediaStream-video-element-change-audio-route.html fast/mediastream/MediaStream-video-element-displays-buffer.html ... (failure)") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/116796 "Found 8 new API test failures: TestWebKitAPI.SiteIsolation.StopsMediaCaptureInRemoteFrame, TestWebKitAPI.WebKit2.CrashGPUProcessAfterApplyingConstraints, TestWebKitAPI.GetUserMedia.ClearRemoteVideoFrameObjectHeapPixelConformerUnderMemoryPressure, TestWebKitAPI.WebKit2.CrashGPUProcessWhileCapturing, TestWebKitAPI.WebKit.InterruptionBetweenSameProcessPages, TestWebKitAPI.WebKit.AutoplayOnVisibilityChange, TestWebKitAPI.WebKit2.WebRTCAndRemoteCommands, TestWebKitAPI.WebKit2.CrashGPUProcessWhileCapturingAndCalling (failure)") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/80068 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/7725e132-a95f-4bb2-bcb6-3b12f6433271) 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/1902 "Found 60 new test failures: imported/w3c/web-platform-tests/css/css-text-decor/text-emphasis-position-auto-001.html imported/w3c/web-platform-tests/css/css-viewport/zoom/iframe-zoom-nested.html imported/w3c/web-platform-tests/css/css-viewport/zoom/zoom-iframe-dynamic.html imported/w3c/web-platform-tests/fetch/security/dangling-markup-mitigation-data-url.tentative.sub.html imported/w3c/web-platform-tests/html/interaction/focus/focus-input-type-switch.html imported/w3c/web-platform-tests/html/rendering/widgets/input-date-no-resize-on-hover.html imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-details-element/details-toggle-source.html imported/w3c/web-platform-tests/html/semantics/popovers/popover-attribute-basic.html imported/w3c/web-platform-tests/mediacapture-insertable-streams/MediaStreamTrackProcessor-maxBufferSize.worker.html imported/w3c/web-platform-tests/mediacapture-insertable-streams/MediaStreamTrackProcessor-stop.worker.html ... (failure)") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/34926 "Found 2 new test failures: fast/mediasession/metadata/artworkdownload-svg.html fast/mediasession/metadata/artworkdownload.html (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/81108 "Built successfully") | | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/110454 "Found 5 new API test failures: TestWebKitAPI.WebKit2.CrashGPUProcessAfterApplyingConstraints, TestWebKitAPI.GetUserMedia.ClearRemoteVideoFrameObjectHeapPixelConformerUnderMemoryPressure, TestWebKitAPI.WebKit2.CrashGPUProcessWhileCapturing, TestWebKitAPI.WebKit.InterruptionBetweenSameProcessPages, TestWebKitAPI.WebKit2.CrashGPUProcessWhileCapturingAndCalling (failure)") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/35431 "Found 60 new test failures: fast/files/blob-stream-error.html fast/images/individual-animation-toggle.html fast/media/template-hidden.html fast/mediasession/metadata/artworkdownload-svg.html fast/mediasession/metadata/artworkdownload.html fast/mediastream/MediaStream-MediaElement-setObject-null.html fast/mediastream/MediaStream-removeTrack-while-playing.html fast/mediastream/MediaStream-video-element-change-audio-route.html fast/mediastream/MediaStream-video-element-displays-buffer.html fast/mediastream/MediaStream-video-element-enter-background.html ... (failure)") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/140327 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/2492 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/2278 "Found 60 new test failures: fast/canvas/webgl/debug-messages-to-console.html fast/dom/document-contentType-data-uri.html fast/forms/color/color-input-swatch-wrapper-appearance-none-inherited.html fast/media/template-hidden.html fast/mediasession/metadata/artworkdownload-svg.html fast/mediasession/metadata/artworkdownload.html fast/mediastream/MediaStream-MediaElement-setObject-null.html fast/mediastream/MediaStream-removeTrack-while-playing.html fast/mediastream/MediaStream-video-element-change-audio-route.html fast/mediastream/MediaStream-video-element-displays-buffer.html ... (failure)") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/107883 "Found 28 new test failures: http/wpt/webcodecs/encoder-task-failing.html http/wpt/webcodecs/videoFrame-colorSpace.html http/wpt/webcodecs/videoFrame-duration.html http/wpt/webcodecs/videoFrame-negative-timestamp.html http/wpt/webcodecs/webcodecs-gc.html imported/w3c/web-platform-tests/webcodecs/full-cycle-test.https.any.html?h264_annexb imported/w3c/web-platform-tests/webcodecs/full-cycle-test.https.any.html?h264_avc imported/w3c/web-platform-tests/webcodecs/full-cycle-test.https.any.html?h265_annexb imported/w3c/web-platform-tests/webcodecs/full-cycle-test.https.any.html?h265_hevc imported/w3c/web-platform-tests/webcodecs/full-cycle-test.https.any.worker.html?h264_annexb ... (failure)") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/2536 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/113138 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/107786 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/1935 "Passed tests") | [❌ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/31581 "Found 60 new test failures: compositing/backing/animate-into-view-with-descendant.html css3/calc/css3-radial-gradients.html css3/viewport-percentage-lengths/vh-resize.html fast/files/blob-stream-error.html fast/media/template-hidden.html fast/mediasession/metadata/artworkdownload-svg.html fast/mediasession/metadata/artworkdownload.html fast/mediastream/MediaStream-MediaElement-setObject-null.html fast/mediastream/MediaStream-removeTrack-while-playing.html fast/mediastream/MediaStream-video-element-change-audio-route.html ... (failure)") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/55451 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/2562 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/65950 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/2380 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/2583 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/2488 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/html/ImageBitmap.cpp:
```diff
@@ -981,6 +981,14 @@ RefPtr<ImageBuffer> ImageBitmap::takeImageBuffer()
+    return m_bitmap->copyNativeImage();
+}
+
+
```
```suggestion
```

youennf commented on Source/WebCore/html/ImageBitmap.cpp:
```diff
@@ -981,6 +981,14 @@ RefPtr<ImageBuffer> ImageBitmap::takeImageBuffer()
+{
+    if (!m_bitmap)
+        return nullptr;
+    return m_bitmap->copyNativeImage();
```
One liner maybe?
Also `m_bitmap` is not const, so we likely want to ref it before calling copyNativeImage.

youennf commented on Source/WebCore/Modules/webcodecs/WebCodecsVideoFrame.cpp:
```diff
@@ -230,44 +230,28 @@ ExceptionOr<Ref<WebCodecsVideoFrame>> WebCodecsVideoFrame::create(ScriptExecutio
-    if (!pixelBuffer)
-        return Exception { ExceptionCode::InvalidStateError,  "Buffer has no frame"_s };
-
-    auto videoFrame = VideoFrame::createFromPixelBuffer(pixelBuffer.releaseNonNull(), { PlatformVideoColorPrimaries::Bt709, PlatformVideoTransferCharacteristics::Iec6196621, PlatformVideoMatrixCoefficients::Rgb, true });
```
Will we get the same colour characteristics in the new code path?
Some WPT tests cover VideoFrame color spaces, hopefully this should be good enough, worth a check though.

kkinnunen-apple committed:
GPUP: WebCodecVideoFrame of ImageBitmap should be constructed via NativeImage
https://bugs.webkit.org/show_bug.cgi?id=301925
rdar://164004657

Reviewed by NOBODY (OOPS!).

WebCodecVideoFrame of ImageBitmap would be constructed with conversion
path ImageBitmap -> ImageBuffer -> PixelBuffer -> VideoFrame.

Use path of ImageBitmap -> ImageBuffer -> NativeImage -> VideoFrame
instead. This gets it closer to the eventual target of ImageBitmap
-> NativeImage -> VideoFrame.

This is work towards being able to hold NativeImage in ImageBitmap,
ImageBitmap -> NativeImage -> VideoFrame. This is performance and
correctness target, where ImageBitmap should not do redundant draws
of image sources. ImageBuffer is a bitmap draw target, where as
NativeImage is a bitmap image source.

Later on, VideoFrame from NativeImage code will be able to construct
the VideoFrame directly out of GPUP-side NativeImage, avoiding
WCP->GPUP transfers.

* Source/WebCore/Modules/webcodecs/WebCodecsVideoFrame.cpp:
(WebCore::WebCodecsVideoFrame::create):
(WebCore::WebCodecsVideoFrame::initializeFrameWithResourceAndSize):
* Source/WebCore/Modules/webcodecs/WebCodecsVideoFrame.h:
* Source/WebCore/html/ImageBitmap.cpp:
(WebCore::ImageBitmap::copyNativeImage):
* Source/WebCore/html/ImageBitmap.h:
* Source/WebCore/platform/VideoFrame.cpp:
(WebCore::VideoFrame::createFromNativeImage):
(WebCore::VideoFrame::fromNativeImage): Deleted.
* Source/WebCore/platform/VideoFrame.h:
* Source/WebCore/platform/graphics/cv/VideoFrameCV.mm:
(WebCore::VideoFrame::createFromNativeImage):
(WebCore::VideoFrame::fromNativeImage): Deleted.
* Source/WebCore/platform/graphics/gstreamer/VideoFrameGStreamer.cpp:
(WebCore::VideoFrame::createFromNativeImage):
(WebCore::VideoFrame::fromNativeImage): Deleted.
